### PR TITLE
Crypt push3

### PR DESCRIPTION
**Pull request by penguin42**
Finally got around to fixing stuff up from Crypt push2 #45 
More lightly tested than previously.
This now uses the unstable sync API to get one time key counts and uses that to decide whether to send more.

### CODE REVIEW

File: matrix-e2e.c [hunk 1]
```diff
@@ -0,0 +1,1767 @@
...
+    hk = (const MatrixHashKeyOlm *)a;
+    return g_str_hash(hk->sender_key) +
+           g_str_hash(hk->sender_id);
+}
+
+static void olm_hash_key_destroy(gpointer k)
+{
+    MatrixHashKeyOlm *ok = k;
+
+    g_free(ok->sender_key);
```
Comment (zamaudio) [inline matrix-e2e.c]: how about setting the keys to random bytes before freeing them?

File: matrix-e2e.c [hunk 2]
```diff
@@ -0,0 +1,1767 @@
...
+    g_free(ok->sender_id);
```
Comment (zamaudio) [inline matrix-e2e.c]: Ditto

File: matrix-e2e.c [hunk 3]
```diff
@@ -0,0 +1,1767 @@
...
+}
+
+static void key_upload_callback(MatrixConnectionData *conn,
+                                gpointer user_data,
+                                struct _JsonNode *json_root,
+                                const char *body,
+                                size_t body_len, const char *content_type);
+
+/* Really clear an area of memory */
+static void clear_mem(volatile char *data, size_t len)
```
Comment (zamaudio) [inline matrix-e2e.c]: I think this function should set random bytes, not set to zeroes.

Comment (penguin42) [reply matrix-e2e.c]: I don't think there's any benefit to using random bytes compared to zeroes; it's not like overwriting magic on a hard drive.

File: matrix-e2e.c (hunk 1 above)
Comment (penguin42) [reply matrix-e2e.c]: ok, I'm calling clear_mem() on sender_key now.

File: matrix-e2e.c (hunk 3 above)
Comment (zamaudio) [reply matrix-e2e.c]: Well to an attacker, receiving zeroes is a sure sign that they have not found a real key, its basically obfuscation at its finest :)

**Pull request closed without merging**